src/main/java/com/grepp/carrierroute/hotel/service/HotelService.java

```diff
@@ -66,25 +73,36 @@ private Hotel findHotelBy(Long id) throws HotelInfoNotFoundedException{
         return hotels;
     }
 
-    private List<HotelRoomDto> getRoomsBy(Hotel hotel, HotelSearchRequestDto requestDto){
-        return hotel.getHotelRooms()
-                .stream()
-                .filter(room -> room.isAvailable(requestDto.getGuestNumber(), requestDto.getNumOfRoom()))
-                .map(converter::convertHotelRoomDto)
-                .collect(Collectors.toList());
-    }
-
     private List<HotelSearchResponseDto> searchHotelsBy(HotelSearchRequestDto requestDto){
         List<HotelSearchResponseDto> hotelSearchResults = new ArrayList<>();
-        List<Hotel> hotels = findHotelsBy(requestDto.getDestinationType(), requestDto.getDestinationName());
 
-        hotels.forEach(hotel -> {
-            List<HotelRoomDto> roomsMatchedByRequest = getRoomsBy(hotel, requestDto);
+        findHotelsBy(requestDto.getDestinationType(), requestDto.getDestinationName()).forEach(hotel -> {
+            List<HotelRoom> roomsMatchedByRequest = findRoomsBy(hotel, requestDto);
+
             if(!roomsMatchedByRequest.isEmpty()) {
-                hotelSearchResults.add(converter.convertHotelSearchResponseDto(hotel, roomsMatchedByRequest));
+                hotelSearchResults.add(converter.convertToHotelSearchResponseDto(hotel, roomsMatchedByRequest));
             }
         });
 
         return hotelSearchResults;
     }
+
+    private List<HotelRoom> findRoomsBy(Hotel hotel, HotelSearchRequestDto requestDto){
+        Map<RoomType, List<HotelRoom>> availableRoomsByType = hotel.getHotelRooms()
+                .stream()
+                .filter(room -> !isRoomBooked(room.getId(), requestDto.getCheckInDate(), requestDto.getCheckOutDate()))
+                .collect(Collectors.groupingBy(HotelRoom::getRoomType));
+
+        return availableRoomsByType.keySet()
+                .stream()
+                .filter(roomType -> availableRoomsByType.get(roomType).size() >= requestDto.getNumOfRoom())
+                .map(availableRoomsByType::get)
+                .flatMap(List::stream)
```

Comment:
```suggestion
                .flatMap(availableRoomsByType::get)
```
스트림 아직 잘 모르지만 이렇게는 안 되나요? ㅎ 안되면 그냥 ㄱㄱ요 ㅎㅎ